File: 13-roman-to-integer/roman-to-integer.kt
class Solution {
    fun romanToInt(s: String): Int {
         if (s.length !in 1..15) throw Exception("String length must be >=1 and <=15")
        val romanCharValues = mapOf(
            "I" to 1,
            "II" to 2,
            "V" to 5,
            "X" to 10,
            "L" to 50,
            "C" to 100,
            "D" to 500,
            "M" to 1000,
            "IV" to 4,
            "IX" to 9,
            "XL" to 40,
            "XC" to 90,
            "CD" to 400,
            "CM" to 900,
        )
        if (!s.all { "$it" in romanCharValues.keys }) throw Exception("we can only accept 'I', 'V', 'X', 'L', 'C', 'D', 'M'")
        val getAsPair = arrayOf("IV", "IX", "XL", "XC", "CD", "CM")
        var i = 0
        var valueOfRoman = 0
        while (i in s.indices) {
            romanCharValues[if (i + 1 in s.indices) {
                if ("${s[i]}${s[i + 1]}" in getAsPair) {
                    "${s[i]}${s[i + 1]}".also { i++ }
                } else "${s[i]}"
            } else "${s[i]}"]?.let {
                valueOfRoman += it
            }
            i++

        }
        return valueOfRoman
    }
}

File: 13-roman-to-integer/roman-to-integer.php
class Solution {

    function romanToInt(string $s): int
{
    // Validate string length
    if (strlen($s) < 1 || strlen($s) > 15) {
        throw new Exception("String length must be >=1 and <=15");
    }

    // Define Roman numeral values using an associative array
    $romanCharValues = [
        "I" => 1,
        "V" => 5,
        "X" => 10,
        "L" => 50,
        "C" => 100,
        "D" => 500,
        "M" => 1000,
        "IV" => 4,
        "IX" => 9,
        "XL" => 40,
        "XC" => 90,
        "CD" => 400,
        "CM" => 900,
    ];

    // Validate character usage
    if (!ctype_alpha($s) || preg_match('/[^IVXLCDM]/', $s)) {
        throw new Exception("We can only accept 'I', 'V', 'X', 'L', 'C', 'D', 'M'");
    }

    $getAsPair = ["IV", "IX", "XL", "XC", "CD", "CM"]; // Special pairs
    $valueOfRoman = 0;

    for ($i = 0; $i < strlen($s); $i++) {
        // Check for special pairs first
        if (isset($s[$i + 1]) && in_array($s[$i] . $s[$i + 1], $getAsPair)) {
            $valueOfRoman += $romanCharValues[$s[$i] . $s[$i + 1]];
            $i++;
        } else {
            $valueOfRoman += $romanCharValues[$s[$i]];
        }
    }

    return $valueOfRoman;
}

}